### PR TITLE
CMake: Fix LLVM detection

### LINKED ISSUE #1777: CMake Error: Target X86 is not in the set of libraries
hi,
I tried compiling master on archlinux (x86_64, cmake 3.26, gcc 13, llvm 15) and the configuration step fails:
```
$ yay -Syu clang llvm zlib ncurses xeus-zmq cmake cppzmq
$ git clone --depth 1 https://github.com/lfortran/lfortran.git
$ cd lfortran
$ cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON -DWITH_LLVM=ON -DWITH_RUNTIME_LIBRARY=ON -DWITH_XEUS=ON -DWITH_ZLIB=ON .
CMake Error at CMakeLists.txt:11 (file):
  file STRINGS file "/tmp/lfortran/version" cannot be read.


-- The C compiler identification is GNU 13.1.1
-- The CXX compiler identification is GNU 13.1.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/sbin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/sbin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found ZLIB: /usr/lib/libz.a (found version "1.2.13")  
-- Performing Test HAVE_FFI_CALL
-- Performing Test HAVE_FFI_CALL - Success
-- Found FFI: /usr/lib/libffi.so  
-- Performing Test Terminfo_LINKABLE
-- Performing Test Terminfo_LINKABLE - Success
-- Found Terminfo: /usr/lib/libtinfo.so  
-- Found zstd: /usr/lib/libzstd.so  
-- Found LibXml2: /usr/lib/libxml2.so (found version "2.10.4") 
-- Found LLVM 15.0.7
-- Using LLVMConfig.cmake in: /usr/lib/cmake/llvm
CMake Error at /usr/lib/cmake/llvm/LLVM-Config.cmake:138 (message):
  Target X86 is not in the set of libraries.
Call Stack (most recent call first):
  /usr/lib/cmake/llvm/LLVM-Config.cmake:263 (llvm_expand_pseudo_components)
  CMakeLists.txt:172 (llvm_map_components_to_libnames)


-- Configuring incomplete, errors occurred!
```
any idea what happens ?

I also tried latest tag with the same result

### PR DESCRIPTION
**Pull request by jschueller**
Seems to fix #1777

### CODE REVIEW

**Comment by jschueller** [comment]
there is also a linking issue I needed to workaround when llvm is compiled to not split libs

**Comment by certik** [comment]
Thanks! I will need to study what `native` means. I would think we need it, but maybe not. What does the
```cmake
    if (NOT TARGET LLVMCore)
        set (llvm_libs "LLVM")
    endif ()
```
part do?

**Comment by jschueller** [comment]
it checks if llvm is compiled into a single libLLVM library or into separate libs libLLVMCore, ...

**Comment by jschueller** [comment]
looks I did not break anything :)

File: CMakeLists.txt [hunk 1]
```diff
@@ -172,6 +172,11 @@ if (WITH_LLVM)
     llvm_map_components_to_libnames(llvm_libs ${LFORTRAN_LLVM_COMPONENTS})
     unset(LFORTRAN_LLVM_COMPONENTS)
 
+    # workaround for https://github.com/llvm/llvm-project/issues/34593
```
Comment (certik) [inline CMakeLists.txt:175]:
```suggestion
    # Workaround for https://github.com/llvm/llvm-project/issues/34593
    # If LLVM is distributed as a single library (the LLVMCore target is missing), we set `llvm_libs` to "LLVM" which links against the single `libLLVM.so` shared library.
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -128,7 +128,7 @@ set(WITH_TARGET_AARCH64 no CACHE BOOL "Enable target AARCH64")
 set(WITH_TARGET_X86 no CACHE BOOL "Enable target X86")
 set(WITH_TARGET_WASM no CACHE BOOL "Enable target WebAssembly")
 if (WITH_LLVM)
-    set(LFORTRAN_LLVM_COMPONENTS core support mcjit orcjit native asmparser asmprinter)
+    set(LFORTRAN_LLVM_COMPONENTS core support mcjit orcjit asmparser asmprinter)
```
Comment (certik) [inline CMakeLists.txt:131]: What is the purpose of this change? Is it still needed? It looks like you override it below by setting `llvm_libs=LLVM` anyway.

What does the `native` component do? I wasn't able to find it easily using google but chatgpt says this: https://chat.openai.com/share/02f016be-45f2-46f9-8709-65da1b982762, so it seems this component does the target machine native code generation. I think we need this component.

Comment (jschueller) [reply CMakeLists.txt:131]: yes, I could completely bypass it
seems its not needed after all since the detection goes fine one other platform on the CI

Comment (certik) [reply CMakeLists.txt:131]: Yes, it seems at least at a CI `native` is not needed.

On your machine, does the build now work with `native` included? If not, what is the error message?

Comment (jschueller) [reply CMakeLists.txt:131]: with "native" cmake failed with : Target X86 is not in the set of libraries
then I had to fixed llvm_libs names for the linking failed

Comment (certik) [reply CMakeLists.txt:131]: I see. Can you try this PR and see if it fixes this issue on your machine: https://github.com/lfortran/lfortran/pull/1791 ?

Comment (jschueller) [reply CMakeLists.txt:131]: yes, it works

Comment (jschueller) [reply CMakeLists.txt:131]: It also seems that linking to "LLVM" works even when LLMCore target is available

**Pull request closed without merging**